api/preProcessor-template.ts

```diff
@@ -0,0 +1,19 @@
+import thing from 'module';
+
+export let sourceExtensionTypes: string[] = [".ts", ".tsx"];
+
+export let outputExtensionTypes: string[] = [".mjs", ".cjs"];
+
+
+// Example of defining "options" for a TypeScript loader
+export interface TypeScriptOptions {
+  tsconfigFile?: string
+}
+
+export function process(source: string | Buffer, options: TypeScriptOptions = {}): string| Buffer {
```

Comment:
I think it may be nice to ensure that the processor is always called with the same kind of `source`, either string or Buffer (same for the return type). That way the implementation doesn't have to be defensive.

Reply:
I see what you mean.  I was basing type off of [`transformSource` hook's](https://nodejs.org/api/esm.html#esm_code_transformsource_code_hook) `source` param.  I think it would simplify implementation, however, to always expect the same type.

I would initially lean `Buffer` for both input and return types in this case, but I will research more about this

Reply:
One thing that’s done for streams in node, is that the API allows setting an encoding. So it could be up to the processor to pick one or the other. I assume most pre-processors would pick string because that’s what Babel/ts/coffee/etc all use.

Reply:
Gotcha, `string` make sense then.  I think I was under the incorrect assumption that `Buffer` would be faster for some reason?  But if `string` is the de facto type for this, then I agree

Reply:
Buffer is faster if it sticks around. Basically the flow is:

1. I/O operations (network, disk) load raw native memory. From node's perspective, that's a `Buffer`. This part is super fast because the data never needs to be converted into V8's data structures.
2. To do string operations on the data (e.g. feed it into a typical parser), the data needs to be converted to a string.
3. To use the data in further I/O operations, it needs to be converted back to a buffer.

The expensive part is every jump from buffer to string or back because it requires running string encoding and copying all the bytes. By using Buffer as the input and output for these functions, it would risk a situation likes this:

1. Load file from disk, it's a buffer.
2. Pass the buffer to the first processor. It needs the data as a string, so it converts it.
3. After the first transform is done, its result is likely a string (because TSC/babel/etc. return strings). To return it, it needs to be converted to a buffer.
4. The buffer is passed into the next processor. It needs the data as a string, so it converts it.
5. [...]

So using a buffer as the API is only faster if all (or at least most) transforms natively use the buffer and never convert it to a string. In the above case, it would be much faster (far fewer conversions) if the buffer had been transformed into a string once at the beginning with everything being strings from then on out.

Reply:
Ah I see, thanks for the breakdown! It given the most likely use cases and the way TS/Babel/etc have done this let's go with `string` for now